vos/vos/commands/vsync.py

```diff
@@ -37,13 +40,265 @@
 
 HOME = os.getenv("HOME", "./")
 
+global_md5_cache = None
+node_dict = {}
+good_dirs = []  # remote directories that exists on the VOSpace service
+
+# placeholder for data local to a thread
+thread_local = threading.local()
+
+
+def compute_md5(filename):
+    """"
+    Computes the md5 of a file and caches the value for subsequent calls
+    """
+    md5 = None
+    if global_md5_cache is not None:
+        md5 = global_md5_cache.get(filename)
+    if md5 is None or md5[2] < os.stat(filename).st_mtime:
+        md5 = md5_cache.MD5Cache.compute_md5(filename,
+                                             block_size=2**19)
+        if global_md5_cache is not None:
+            stat = os.stat(filename)
+            global_md5_cache.update(filename, md5, stat.st_size,
+                                    stat.st_mtime)
+    else:
+        md5 = md5[0]
+    return md5
+
+
+def get_client(certfile, token):
+    """
+    Returns a VOS client instance for each thread. VOS Client uses requests
+    session which is not thread safe hence creating one instance of this class
+    for each thread
+    :param certfile:
+    :param token:
+    :return: vos.Client
+    """
+    if not hasattr(thread_local, "client"):
+        thread_local.client = vos.Client(vospace_certfile=certfile,
+                                         vospace_token=token)
+    return thread_local.client
+
+
+class TransferReport:
+    """
+    Report of a job.
+    """
+    def __init__(self):
+        self.bytes_sent = 0
+        self.files_sent = 0
+        self.bytes_skipped = 0
+        self.files_skipped = 0
+        self.files_erred = 0
+
+    def __eq__(self, other):
+        return (self.bytes_sent == other.bytes_sent) and \
+               (self.files_sent == other.files_sent) and \
+               (self.bytes_skipped == other.bytes_skipped) and \
+               (self.files_skipped == other.files_skipped) and \
+               (self.files_erred == other.files_erred)
+
+
+def execute(src, dest, opt):
+    """
+    Transfer a file from source to destination
+    :param src: local path to file to transfer
+    :param dest: vospace location
+    :param opt: command line parameters
+    :return: TransferReport()
+    """
+    result = TransferReport()
```

Comment:
Because I'm still trying to figure out how `thread_local` works, why doesn't the `TransferReport` instance also have to be in `thread_local`? Thanks.

Reply:
There's a certain overhead with creating a `vos.Client` (with it's underlying `request.Session`) so the goal is to minimize the number of instances that created. Ideally, I would have liked to re-use the instance created in the main thread (l347) but according the the article, `request.Session` is not thread safe so the next best thing is to create one instance for each thread which is what `thread_local` insures. You can think of `thread_local` as a memory space for a thread.
I could have also cumulated the transfer reports in `thread_local` for collecting reports for each thread but the overhead of creating a new `TransferReport` is so low (just allocate a bit of memory) that was not worth it.